WooCommerce/WooCommerceTests/Model/MarkOrderAsReadUseCaseTests.swift

```diff
@@ -0,0 +1,189 @@
+
+        if let note = sampleNote(read: false), let orderID = note.meta.identifier(forKey: .order) {
+            setupStoreManagerReceivingNotificationActions(for: note, noteStore: noteStore)
+            noteStore.updateLocalNotes(with: [note]) {
+                XCTAssertEqual(self.viewStorage.countObjects(ofType: Storage.Note.self), 1)
+                Task {
+                    let result = await MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded(stores: self.storesManager, noteID: note.noteID, orderID: orderID)
+                    switch result {
+                    case .success(let markedNote):
+                        XCTAssertEqual(note.noteID, markedNote.noteID)
+                        let storageNote = self.viewStorage.loadNotification(noteID: markedNote.noteID)
+                        XCTAssertEqual(storageNote?.read, true)
+                        expectation.fulfill()
+                    case .failure(let error):
+                        XCTFail(error.localizedDescription)
+                    }
+                }
+            }
+        } else {
+            XCTFail()
+        }
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    @MainActor
+    func test_markOrderNoteAsReadIfNeeded_with_stores_alreadyReadNote() {
+        let expectation = self.expectation(description: "Mark order as read with stores")
+        let noteStore = NotificationStore(dispatcher: dispatcher, storageManager: storageManager, network: network)
+
+        if let note = sampleNote(read: true), let orderID = note.meta.identifier(forKey: .order) {
+            setupStoreManagerReceivingNotificationActions(for: note, noteStore: noteStore)
+            noteStore.updateLocalNotes(with: [note]) {
+                XCTAssertEqual(self.viewStorage.countObjects(ofType: Storage.Note.self), 1)
+                Task {
+                    let result = await MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded(stores: self.storesManager, noteID: note.noteID, orderID: orderID)
+                    switch result {
+                    case .success:
+                        XCTFail("Note was already read, it should not be marked as read again.")
+                    case .failure(let error):
+                        if case MarkOrderAsReadUseCase.Error.noNeedToMarkAsRead = error {
+                            expectation.fulfill()
+                        } else {
+                            XCTFail("Got wrong error \(error.localizedDescription)")
+                        }
+                    }
+                }
+            }
+        } else {
+            XCTFail()
+        }
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+    @MainActor
+    func test_markOrderNoteAsReadIfNeeded_with_network_unreadNote() async {
+        let expectation = self.expectation(description: "Mark order as read with network")
```

Comment:
This is just a style preference, but I find this kind of expectation-based test a bit harder to read than one using e.g. `waitForAsync`. For example:

```swift
@MainActor
func test_markOrderNoteAsReadIfNeeded_with_network_unreadNote() async {
    guard let note = self.sampleNote(read: false), let orderID = note.meta.identifier(forKey: .order) else {
        return XCTFail()
    }

    let result = await waitForAsync { promise in
        self.network.simulateResponse(requestUrlSuffix: "notifications", filename: "notifications-load-all")
        self.network.simulateResponse(requestUrlSuffix: "notifications/read", filename: "generic_success")
        promise(await MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded(network: self.network, noteID: note.noteID, orderID: orderID))
    }

    switch result {
    case .success(let markedNoteID):
        XCTAssertEqual(note.noteID, markedNoteID)
    case .failure(let error):
        XCTFail(error.localizedDescription)
    }
}
```

Anyway, no need to change the test; I just wanted to offer that as an alternative.

Reply:
Picking up from @rachelmcr 's observation, I believe it should be possible to do away entirely with custom code do manage the async code.

```swift
    @MainActor
    func test_markOrderNoteAsReadIfNeeded_with_network_unreadNote() async throws {
        let note = try XCTUnwrap(sampleNote(read: false))
        let orderID = try XCTUnwrap(note.meta.identifier(forKey: .order))

        network.simulateResponse(requestUrlSuffix: "notifications", filename: "notifications-load-all")
        network.simulateResponse(requestUrlSuffix: "notifications/read", filename: "generic_success")

        let result = await MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded(network: network, noteID: note.noteID, orderID: orderID)

        switch result {
        case .success(let markedNoteID):
          XCTAssertEqual(note.noteID, markedNoteID)
        case .failure(let error):
          XCTFail(error.localizedDescription)
        }
    }
```

This code seems better to me because:

- Uses no custom code, just Swift's `async / await` and XCTest assertions
- Has a one level indentation 
- By removing closures, also removes the cumbersome `self.` accesses

_Works on my machine_™

![image](https://github.com/user-attachments/assets/6f1cd8f5-b0e3-482e-bc31-1527d726a834)